### PR TITLE
feat(plugin): add dpop plugin for RFC 9449 proof-of-possession

### LINKED ISSUE #11219: feat: support DPoP in OpenID/OAuth2 plugin
### Description

Demonstrating Proof of Possession (DPoP, RFC9449) is a new proposed standard (https://datatracker.ietf.org/doc/html/rfc9449) for OAuth2 that constraints who can use the access token for when calling an OAuth2-protected endpoint. It's an important standard for improving the security of OAuth2 tokens when migrating to from mTLS to OAuth2 deployments.

It works by the client sharing a JWK thumbprint when requesting an access token. The client then signs a DPoP proof (containing request URL and HTTP method) and provides it as HTTP header when accessing the OAuth2-protected API.
The resource server (or APISIX in this case) then uses the `cnf` claim from the OAuth2 Token Introspection response to verify the proof signature. It can then reassure the OAuth2 access token wasn't stolen by a MITM, but is in fact used by the very same party that requested the token. Usage is not enforced by the server, but up to the OAuth2 client to enable.

Are there plans to support DPoP?

### PR DESCRIPTION
**Pull request by Alnyli07**
## Description

Add `dpop` plugin for [RFC 9449 DPoP (Demonstrating Proof of Possession)](https://datatracker.ietf.org/doc/html/rfc9449) validation. This plugin enforces sender-constrained access tokens at the API gateway level, preventing token theft and replay attacks with zero changes to upstream services.

Fixes https://github.com/apache/apisix/issues/11219

### Why a Separate Plugin?

DPoP is a **proof-of-possession layer**, not an authentication mechanism. It complements existing auth plugins (`openid-connect`, `jwt-auth`, `multi-auth`) rather than replacing them. It runs at priority **2601** to rewrite `Authorization: DPoP` → `Authorization: Bearer` before downstream auth plugins process the token.

### Features

- Full RFC 9449 DPoP proof validation (`typ`, `alg`, `htm`, `htu`, `iat`, `ath`, `jti`)
- DPoP proof signature verification (ES256/384/512, RS256/384/512, PS256/384/512)
- JWK Thumbprint binding (`cnf.jkt`) per RFC 7638
- Access token signature verification via JWKS (OIDC discovery or direct URI)
- Token introspection fallback (RFC 7662) for opaque tokens
- Distributed JTI replay protection (memory / Redis / Infinispan, configurable fallback strategy)
- Bearer downgrade detection
- `DPoP` → `Bearer` header rewrite + `DPoP-Thumbprint` header injection
- Selective enforcement via `uri_allow`
- Structured audit logging with OpenTelemetry trace_id correlation

A Go WASM version (proxy-wasm-go-sdk v0.24.0) is also available at [Keymate-io/keymate-apisix-dpop-plugin](https://github.com/Keymate-io/keymate-apisix-dpop-plugin).

### Files Changed

| File | Description |
|------|-------------|
| `apisix/plugins/dpop.lua` | Plugin implementation (1384 LOC) |
| `conf/config.yaml.example` | Plugin registration (priority 2601) |
| `docs/en/latest/plugins/dpop.md` | English documentation |
| `t/plugin/dpop.t` | 15 Test::Nginx test cases |

### Tested With

- Apache APISIX 3.11.0
- Keycloak 26.0 (any RFC 9449-compliant IdP)
- 58 E2E checks (k6), 49 Postman requests in our upstream repo
- Dependencies: only libraries already in APISIX (`resty.openssl`, `resty.http`, `resty.redis`, `resty.lrucache`)

## Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible

### CODE REVIEW

File: apisix/plugins/dpop.lua [hunk 1]
```diff
@@ -0,0 +1,1384 @@
...
+    properties = {
+        allowed_algs = {
+            type = "array",
+            items = {
+                type = "string",
+                enum = {
+                    "ES256", "ES384", "ES512",
+                    "RS256", "RS384", "RS512",
+                    "PS256", "PS384", "PS512",
+                },
```
Comment (Baoyuantop) [inline apisix/plugins/dpop.lua:111]: But does verify_dpop_proof_signature() only implement ES256 and RS256/RS384/RS512, a total of 4?

File: apisix/plugins/dpop.lua [hunk 2]
```diff
@@ -0,0 +1,1384 @@
...
+    local now = ngx.time()
+    local cached = _discovery_cache[conf.discovery]
+    local ttl = conf.jwks_cache_ttl or 86400
+    if cached and (now - cached.fetched_at) < ttl then
+        return cached.jwks_uri
+    end
+
+    local httpc = http.new()
+    httpc:set_timeout(5000)
+    local res, err = httpc:request_uri(conf.discovery, { method = "GET", ssl_verify = false })
```
Comment (Baoyuantop) [inline apisix/plugins/dpop.lua]: It is recommended to add the ssl_verify configuration item (default is true), or at least refer to the approach of the openid-connect plugin to allow users to control whether to skip TLS verification.

File: apisix/plugins/dpop.lua (hunk 1 above)
Comment (Alnyli07) [reply apisix/plugins/dpop.lua:111]: Thanks for flagging this.

Our primary test environment uses Keycloak which defaults to ES256 for DPoP                                                                      
proofs, and RS256/384/512 for access token signing — so those were the                                                                           
algorithms we had full E2E coverage for. The schema was populated from the                                                                       
full RFC 9449 recommended algorithm list for completeness, but the                                                                               
verification implementation didn't keep pace with the schema. That's on us.                                                                      
                                                                                                                                                  
We'll add ES384, ES512, PS256, PS384, and PS512 support — `resty.openssl.pkey`                                                                   
already handles these, it's mainly a matter of mapping the correct hash                                                                          
algorithm and padding parameters.                                                                                                                
                                                                                                                                                  
While we're at it — are there any other algorithms you'd like to see                                                                             
supported, or is the current RFC 9449 set (EC + RSA + RSA-PSS families)                                                                          
sufficient?

File: apisix/plugins/dpop.lua (hunk 2 above)
Comment (Alnyli07) [reply apisix/plugins/dpop.lua]: Valid point. This was left from our development setup and should definitely be configurable.

We'll add an `ssl_verify` attribute (boolean, default `true`) to the schema
and propagate it to all HTTP calls (discovery, JWKS, introspection). 
Consistent with how `openid-connect` handles it.

Comment (Alnyli07) [reply apisix/plugins/dpop.lua]: Landed in **7b6c419**.  The hardcoded `false` was a leftover from local
development.

The schema gains an `ssl_verify` boolean (default `true`), propagated
to every outbound HTTP call the plugin issues:

- `resolve_jwks_uri` (OIDC discovery fetch)
- `fetch_jwks` (JWKS fetch; the function signature now takes
  `ssl_verify` so the call site can pass `conf.ssl_verify` through)
- `jti_check` (Infinispan replay-cache backend, including the
  digest-auth retry path)
- `call_introspection` (RFC 7662 introspection endpoint)

Default `true` keeps existing deployments behaving the same as before;
the opt-out is only available when operators set it explicitly. Same
shape as `openid-connect`'s `ssl_verify`.

File: apisix/plugins/dpop.lua (hunk 1 above)
Comment (Alnyli07) [reply apisix/plugins/dpop.lua:111]: **75f2d87** : the verification path now dispatches all nine asymmetric algs declared
in the schema: ES256/384/512, RS256/384/512, PS256/384/512.

Two implementation notes:

1. ECDSA proofs carry signatures in raw R||S (JWS Compact). Converting
   to the DER form OpenSSL expects needs long-form SEQUENCE length once
   R+S exceeds 127 bytes (every ES512 signature). The `raw_ecdsa_to_der`
   helper now emits long-form (`0x81`/`0x82`).
2. RSA-PSS verification under OpenSSL 3's provider API requires
   `RSA_PKCS1_PSS_PADDING` to be set before any saltlen/MGF1 parameter.
   The plugin now passes padding via the dedicated argument and uses
   the named `pss_saltlen` setter (digest length per RFC 7518 §3.5),
   so parameters are applied in the order the provider enforces.

**0028e83** : TEST 31 (ES512) and TEST 32-35 (RS384/RS512/PS384/PS512) cover the
newly-added algorithms. The full suite has 35 cases / 105 assertions
and passes against the standard upstream test runner.

**Comment by Alnyli07** [comment]
@Baoyuantop both threads above are now addressed. Brief summary of what
landed on top of the initial PR:

- `ssl_verify` config option (default `true`), propagated to every
  outbound HTTP call the plugin issues.
- Verification path now covers all nine RFC 7518 asymmetric algorithms
  declared in the schema (ES256/384/512, RS256/384/512, PS256/384/512),
  with the long-form DER and OpenSSL 3 PSS adjustments noted in the
  thread.
- Test::Nginx suite covering schema validation, route enforcement, the
  full crypto flow per algorithm, and proof-claim negatives. 35 cases /
  105 assertions, all passing.

**Comment by Alnyli07** [comment]
Hi everyone, 

Just bumping this PR. All feedback from @Baoyuantop has been fully addressed, and the 35 E2E test cases are passing cleanly.

Could we get additional reviews to help meet the 3-approval requirement? Thanks! 🙏

**Comment by Baoyuantop** [comment]
Thanks for continuing to improve this DPoP plugin. The current revision does address the two earlier review points: `ssl_verify` is now configurable and propagated to outbound discovery/JWKS calls, and proof signature verification now covers the ES/RS/PS algorithm families with additional tests.

However, I do not think this is mergeable yet. There are a few blocking issues that should be fixed first:

1. A malformed proof without `alg` can hit a Lua runtime error. `verify_dpop_proof_signature` eventually concatenates the nil `alg` value when no algorithm branch matches, and `validate_proof` also calls `h.alg:sub(...)` directly. This kind of input should consistently return `invalid_dpop_proof`, not fall into a 500/error path. Please add explicit header structure validation and negative tests for missing/non-string `alg`.

2. `require_nonce` is accepted by the schema, while the documentation says it is “not yet implemented”, and the implementation does not validate a DPoP nonce. A security option should not be accepted as a no-op. Please either implement the nonce semantics with tests, or remove the option for now.

3. With the default `strict_htu=false`, the plugin only compares the path and ignores scheme/host/port. This compatibility mode can weaken the URI binding semantics from RFC 9449. Please re-evaluate the default behavior, or at least document it clearly and add tests for strict/full-URL and path-only boundaries.

Please address these points and request another review.

**Comment by Alnyli07** [comment]
Thanks for the review again @Baoyuantop. All three points addressed:

1. Malformed proof / missing alg: Added header/payload structure validation that runs before signature verification and validate_proof, rejecting a missing/non-string alg (and malformed header/payload/jwk) with invalid_dpop_proof. The nil-concat and h.alg:sub() paths are now unreachable. Added negative tests for missing alg, non-string alg, and non-object header/payload (each expects 401, no error log).

2. **require_nonce**:  Removed from schema and docs, since the nonce semantics aren't implemented (no longer a no-op option).

3. **strict_htu** default Kept false because strict_htu=true requires public_base_url, so defaulting to true would break existing routes. The path-only behavior and its weakening of RFC 9449 §4.3 URI binding is now documented explicitly, with tests for full-URL match (200), host/scheme mismatch (401), and path-only default (200). Happy to flip the default if you'd prefer secure-by-default.

Ready for another look.

**Comment by Alnyli07** [comment]
Hi @Baoyuantop and everyone! 👋 All review feedback has been addressed and all tests are passing. Could we get a couple more reviews to help move this forward? Thanks! 🙏
